frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -44,187 +43,251 @@
+    int pendingTaskNumber = statusManager.getUnassociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
+    if (requestManager.getTaskRoles().get(taskRoleName).getUseTheSamePorts()) {
+      reUsePorts = statusManager.getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && previousRequestedPorts.containsKey(taskRoleName)) {
+        reUsePorts = previousRequestedPorts.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        previousRequestedPorts.remove(taskRoleName);
       }
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        if (configuredNodes != null) {
-          if (!configuredNodes.containsKey(nodeHost)) {
-            LOGGER.logDebug(rejectedLogPrefix +
-                    "Node is not found in configured Nodes: Request NodeGpuType: [%s]",
-                requestNodeGpuType);
-            continue;
-          }
-
-          List<String> requestNodeGpuTypes = Arrays.asList(requestNodeGpuType.split(","));
-          String availableNodeGpuType = configuredNodes.get(nodeHost).getGpuType();
-          if (!requestNodeGpuTypes.contains(availableNodeGpuType)) {
-            LOGGER.logDebug(rejectedLogPrefix +
-                    "NodeGpuType does not match: Request NodeGpuType: [%s], Available NodeGpuType: [%s]",
-                requestNodeGpuType, availableNodeGpuType);
-            continue;
-          }
-        } else {
-          LOGGER.logWarning(logPrefix +
-                  "Configured Nodes is not found in ClusterConfiguration: Ignore Request NodeGpuType: [%s]",
-              requestNodeGpuType);
-        }
+    }
+    SelectionResult result = select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, reUsePorts);
+
+    // This pendingTaskNumber includes current task, when pendingTaskNumber == 1 means current task is the last task.
+    // If there has other task pending, push current port to previousRequestedPorts for next task re-use.
+    // reusePortsTimes time is used to avoid pending taskTaskNumber not decrease when previous jobs time out while next job not started.
+    if (pendingTaskNumber > 1) {
+      if(reusePortsTimes == 0) {
+        reusePortsTimes = pendingTaskNumber;
       }
-
-      // Test Resource
-      ResourceDescriptor availableResource = node.getAvailableResource();
-      if (!ResourceDescriptor.fitsIn(requestResource, availableResource)) {
-        LOGGER.logDebug(rejectedLogPrefix +
-                "Resource does not fit in: Request Resource: [%s], Available Resource: [%s]",
-            requestResource, availableResource);
-        continue;
+      if(reusePortsTimes > 1) {
+        previousRequestedPorts.put(taskRoleName, result.getOptimizedResource().getPortRanges());
       }
```

Comment:
This logic should be in

if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && previousRequestedPorts.containsKey(taskRoleName)) {